components/automate-ui/src/app/modules/infra-proxy/infra-role-details/infra-role-details.component.ts

```diff
@@ -0,0 +1,216 @@
+  public childNodes: Node<ChildList>[] = [];
+  private isDestroyed = new Subject<boolean>();
+  arrayOfNodesTree: Node<ChildList>[];
+  roleDetailsLoading = true;
+  treeOptions: Options<ChildList> = {
+    capitalisedHeader: true
+  };
+
+  public attributes: RoleAttributes = new RoleAttributes({
+    default_attributes: '',
+    override_attributes: ''
+  });
+
+  public selectedAttrs: any;
+  public selected_level = 'all';
+
+  // precedence levels
+  public default_attributes = 'default_attributes';
+  public override_attributes = 'override_attributes';
+  public all = 'all';
+
+  @ViewChild(Treeable, { static: true })
+  tree: Treeable;
+
+  constructor(
+    private store: Store<NgrxStateAtom>,
+    private router: Router,
+    private layoutFacade: LayoutFacadeService
+  ) { }
+
+  ngOnInit() {
+    this.arrayOfNodesTree = [];
+    this.childNodes = [];
+    this.idList = [];
+    this.layoutFacade.showSidebar(Sidebar.Infrastructure);
+    this.store.select(routeURL).pipe(takeUntil(this.isDestroyed))
+      .subscribe((url: string) => {
+        this.url = url;
+        const [, fragment] = url.split('#');
+        this.tabValue = (fragment === 'attributes') ? 'attributes' : 'runList';
+      });
+
+    combineLatest([
+      this.store.select(routeParams).pipe(pluck('id'), filter(identity)),
+      this.store.select(routeParams).pipe(pluck('orgid'), filter(identity)),
+      this.store.select(routeParams).pipe(pluck('name'), filter(identity))
+    ]).pipe(
+      takeUntil(this.isDestroyed)
+    ).subscribe(([server_id, org_id, name]: string[]) => {
+      this.serverId = server_id;
+      this.OrgId = org_id;
+      this.name = name;
+      this.store.dispatch(new GetRole({
+        server_id: server_id, org_id: org_id, name: name
+      }));
+    });
+
+    this.store.select(infaRoleFromRoute).pipe(
+      filter(identity),
+      takeUntil(this.isDestroyed)
+    ).subscribe(role => {
+      this.show = true;
+      this.role = role;
+      this.expandedList = role.expanded_run_list;
+      this.runList = this.role.run_list;
+      this.idList = [];
+      this.attributes = new RoleAttributes(this.role);
+      if (this.expandedList && this.expandedList.length) {
+        this.show = true;
+        for (let i = 0; i < this.expandedList.length; i++) {
+          this.idList.push(
+            this.expandedList[i].id
+          );
+        }
+        if (this.tabValue === 'runList') {
+          this.treeNodes(this.expandedList, this.env_id);
+        }
+
+      } else {
+        this.show = false;
+      }
+      setTimeout(() => this.filter(this.selected_level), 10);
+      this.roleDetailsLoading = false;
+    });
+  }
+
+  // set selected item to selected_level
+  filter(precedence_level: string): void {
+    this.selected_level = precedence_level;
+    this.selectedAttrs = this.retrieve(precedence_level);
+  }
+
+  // retrieve attributes based on their level of precedence
+  retrieve(level: string): any {
```

Comment:
Why is this returning a string everywhere except for the default when it returns an object?
So would prefer to see this like this if possible:
```suggestion
  retrieve(level: string): string {
```
If not, then use a union type (https://www.typescriptlang.org/docs/handbook/advanced-types.html#union-types):
```suggestion
  retrieve(level: string): string | Object {
```